### PR TITLE
feat(FEC-11387): move from base.ini to appVersions configuration in admin

### PR DESCRIPTION
**Pull request by Yuvalke**

### CODE REVIEW

**Review by OrenMe — APPROVED** [review]

**Comment by OrenMe** [comment]
need to create app_versions in prod env

File: alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php [hunk 1]
```diff
@@ -158,14 +158,24 @@ public function execute ( )
...
-		$this->content_uiconf_studio_v2 = (is_array($this->content_uiconfs_studio_v2) && reset($this->content_uiconfs_studio_v2)) ? reset($this->content_uiconfs_studio_v2) : null;
 
-		$this->studioV3UiConf = uiConfPeer::getUiconfByTagAndVersion(self::STUDIO_V3_TAG, kConf::get("studio_v3_version"));
-		$this->content_uiconfs_studio_v3 = isset($this->studioV3UiConf) ? array_values($this->studioV3UiConf) : null;
-		$this->content_uiconf_studio_v3 = (is_array($this->content_uiconfs_studio_v3) && reset($this->content_uiconfs_studio_v3)) ? reset($this->content_uiconfs_studio_v3) : null;
+		$embedPlaykitConf = kConf::getMap(kConfMapNames::APP_VERSIONS);
+		$playerApps = isset($embedPlaykitConf['playerApps']) ? $embedPlaykitConf['playerApps'] : null;
+		$html5_version = isset($playerApps['html5_version']) ? $playerApps['html5_version'] : null;
+		$studio_version = isset($playerApps['studio_version']) ? $playerApps['studio_version'] : null;
+		$studio_v3_version = isset($playerApps['studio_version']) ? $playerApps['studio_v3_version'] : null;
```
Comment (yairans) [inline alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php:166]:
```suggestion
		$studio_v3_version = isset($playerApps['studio_v3_version']) ? $playerApps['studio_v3_version'] : null;
```

**Review by yairans — CHANGES_REQUESTED** [review]

**Review by yairans — APPROVED** [review]

File: alpha/apps/kaltura/modules/extwidget/actions/embedIframeJsAction.class.php [hunk 2]
```diff
@@ -77,7 +77,9 @@ public function execute()
 		{
 			if (!$iframeEmbed)
 				$host = "$protocol://". kConf::get('html5lib_host') ."/";
-			$html5_version = kConf::get('html5_version');
+			$embedPlaykitConf = kConf::getMap(kConfMapNames::APP_VERSIONS);
```
Comment (MosheMaorKaltura) [inline alpha/apps/kaltura/modules/extwidget/actions/embedIframeJsAction.class.php:80]: Better approach in one line - 
kConf::getArrayValue('html5_version','playerApps',kConfMapNames::APP_VERSIONS,null);

Signature - getArrayValue($paramName, $sectionName, $mapName = 'local', $defaultValue = false)

File: alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php [hunk 3]
```diff
@@ -122,18 +122,20 @@ public function execute()
...
 		if(!$uiConf)
 			KExternalErrors::dieError(KExternalErrors::UI_CONF_NOT_FOUND);
-			
+
 		$partner_host = myPartnerUtils::getHost($partner_id);
 		$partner_cdnHost = myPartnerUtils::getCdnHost($partner_id);
-		
-		$html5_version = kConf::get('html5_version');
+
+		$embedPlaykitConf = kConf::getMap(kConfMapNames::APP_VERSIONS);
```
Comment (MosheMaorKaltura) [inline alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php:136]: Better approach in one line -
kConf::getArrayValue('html5_version','playerApps',kConfMapNames::APP_VERSIONS,null);

Signature - getArrayValue($paramName, $sectionName, $mapName = 'local', $defaultValue = false)

File: alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php [hunk 4]
```diff
@@ -158,14 +158,24 @@ public function execute ( )
...
-		$this->content_uiconfs_studio_v2 = isset($this->studioUiConf) ? array_values($this->studioUiConf) : null;
-		$this->content_uiconf_studio_v2 = (is_array($this->content_uiconfs_studio_v2) && reset($this->content_uiconfs_studio_v2)) ? reset($this->content_uiconfs_studio_v2) : null;
 
-		$this->studioV3UiConf = uiConfPeer::getUiconfByTagAndVersion(self::STUDIO_V3_TAG, kConf::get("studio_v3_version"));
-		$this->content_uiconfs_studio_v3 = isset($this->studioV3UiConf) ? array_values($this->studioV3UiConf) : null;
-		$this->content_uiconf_studio_v3 = (is_array($this->content_uiconfs_studio_v3) && reset($this->content_uiconfs_studio_v3)) ? reset($this->content_uiconfs_studio_v3) : null;
+		$embedPlaykitConf = kConf::getMap(kConfMapNames::APP_VERSIONS);
+		$playerApps = isset($embedPlaykitConf['playerApps']) ? $embedPlaykitConf['playerApps'] : null;
+		$html5_version = isset($playerApps['html5_version']) ? $playerApps['html5_version'] : null;
+		$studio_version = isset($playerApps['studio_version']) ? $playerApps['studio_version'] : null;
```
Comment (MosheMaorKaltura) [inline alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php:165]: kConf::getArrayValue('studio_version','playerApps',kConfMapNames::APP_VERSIONS,null);

File: alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php [hunk 5]
```diff
@@ -158,14 +158,24 @@ public function execute ( )
...
+		$studio_v3_version = isset($playerApps['studio_v3_version']) ? $playerApps['studio_v3_version'] : null;
```
Comment (MosheMaorKaltura) [inline alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php:166]: kConf::getArrayValue('studio_v3_version','playerApps',kConfMapNames::APP_VERSIONS,null);

File: alpha/apps/kaltura/modules/kmcng/actions/kmcngAction.class.php [hunk 6]
```diff
@@ -108,22 +108,27 @@ private function initConfig($deployUrl, $kmcngParams, $enforceSecureProtocol, $r
 
 
 		$studio = null;
-		if (kConf::hasParam("studio_version") && kConf::hasParam("html5_version"))
+		$embedPlaykitConf = kConf::getMap(kConfMapNames::APP_VERSIONS);
```
Comment (MosheMaorKaltura) [inline alpha/apps/kaltura/modules/kmcng/actions/kmcngAction.class.php:111]: kConf::getArrayValue('studio_v3_version','playerApps',kConfMapNames::APP_VERSIONS,null);

**Review by MosheMaorKaltura — COMMENTED** [review]
See my note about the correct method to use the kConf API

File: alpha/config/kConfMapNames.php [hunk 7]
```diff
@@ -6,6 +6,7 @@ class kConfMapNames
 {
 	const CACHE_VERSIONS = 'cache_versions';
 	const EMBED_PLAYKIT = 'embed_playkit';
+	const APP_VERSIONS = 'app_versions';
```
Comment (yairans) [inline alpha/config/kConfMapNames.php:9]:
```suggestion
	const APP_VERSIONS = 'appVersions';
```
no?

File: alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php [hunk 8]
```diff
@@ -158,14 +158,22 @@ public function execute ( )
 
 		$this->content_uiconds_clipapp_kdp = kmcUtils::find_confs_by_usage_tag($kmcGeneralTemplateUiConf, "kmc_kdpClipApp", false, $kmcGeneralUiConf);
 		$this->content_uiconds_clipapp_kclip = kmcUtils::find_confs_by_usage_tag($kmcGeneralTemplateUiConf, "kmc_kClipClipApp", false, $kmcGeneralUiConf);
-		
-		$this->studioUiConf = uiConfPeer::getUiconfByTagAndVersion(self::HTML5_STUDIO_TAG, kConf::get("studio_version"));
```
Comment (inbal-ben-david) [inline alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php:162]: indetation

File: alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php [hunk 9]
```diff
@@ -158,14 +158,22 @@ public function execute ( )
...
-		$this->content_uiconf_studio_v2 = (is_array($this->content_uiconfs_studio_v2) && reset($this->content_uiconfs_studio_v2)) ? reset($this->content_uiconfs_studio_v2) : null;
 
-		$this->studioV3UiConf = uiConfPeer::getUiconfByTagAndVersion(self::STUDIO_V3_TAG, kConf::get("studio_v3_version"));
-		$this->content_uiconfs_studio_v3 = isset($this->studioV3UiConf) ? array_values($this->studioV3UiConf) : null;
-		$this->content_uiconf_studio_v3 = (is_array($this->content_uiconfs_studio_v3) && reset($this->content_uiconfs_studio_v3)) ? reset($this->content_uiconfs_studio_v3) : null;
+        $html5_version = kConf::getArrayValue('html5_version','playerApps',kConfMapNames::APP_VERSIONS,null);
+        $studio_version = kConf::getArrayValue('studio_version','playerApps',kConfMapNames::APP_VERSIONS,null);
+        $studio_v3_version = kConf::getArrayValue('studio_v3_version','playerApps',kConfMapNames::APP_VERSIONS,null);
+
+		if (!is_null($studio_version)) {
```
Comment (inbal-ben-david) [inline alpha/apps/kaltura/modules/kmc/actions/kmc4Action.class.php:166]: if
{
}

File: alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php [hunk 10]
```diff
@@ -122,18 +122,18 @@ public function execute()
...
 		if(!$uiConf)
 			KExternalErrors::dieError(KExternalErrors::UI_CONF_NOT_FOUND);
-			
+
 		$partner_host = myPartnerUtils::getHost($partner_id);
 		$partner_cdnHost = myPartnerUtils::getCdnHost($partner_id);
-		
-		$html5_version = kConf::get('html5_version');
+
+		$html5_version = kConf::getArrayValue('html5_version','playerApps',kConfMapNames::APP_VERSIONS,null);
```
Comment (inbal-ben-david) [inline alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php:136]: on all places you call to getArrayValue add spaces after each ','

**Review by inbal-ben-david — COMMENTED** [review]
- add release notes explaining you moved the config from base.ini to appVersions
- create new template map as a fallback named appVersions.template.ini under configurations folder with templated values

File: alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php (hunk 10 above)
Comment (OrenMe) [reply alpha/apps/kaltura/modules/extwidget/actions/embedIframeAction.class.php:136]: @inbal-ben-david may I suggest using a formatting tool that will automatically format the code according to standards
https://github.com/prettier/plugin-php - highly recommended!

**Comment by yairans** [comment]
@inbal-ben-david what do you mean template map as a fallback? Does the server know to get this map in runtime if the appVersions map doesn't exist?

**Comment by yairans** [comment]
@inbal-ben-david what do you mean template map as a fallback? Does the server know to get this map in runtime if the appVersions map doesn't exist?

**Comment by yairans** [comment]
@jessp01 please see this PR. it's a breaking change to the CE.

File: release-notes.md [hunk 11]
```diff
@@ -1,4 +1,18 @@
 # Quasar-17.15.0 #
+## Move V2 player version and studio versions from base.ini to dedicated map ##
```
Comment (inbal-ben-david) [inline release-notes.md:2]: move release notes to Quasar-17.16.0

File: alpha/apps/kaltura/lib/KExternalErrors.class.php [hunk 12]
```diff
@@ -129,6 +129,7 @@ class KExternalErrors
 		self::MISSING_LIVE_CONFIGURATION => "Missing live configuration",
 		self::DATABASE_ERROR => 'Database error',
 		self::INVALID_AUTH_HEADER => 'Invalid auth header',
+		self::VERSION_NOT_FOUNDVERSION_NOT_FOUND => "The [%s] version was not found"
```
[the remark this answers is not in the document]
Comment (yairans) [reply alpha/apps/kaltura/lib/KExternalErrors.class.php]: oops. thanks 👍

**Comment by jessp01** [comment]
Hi @yairans,

This is fine. I will replace the  tokens in configurations/appVersions.template.ini during the configuration phase and create configurations/appVersions.ini

Anything else that needs doing here?

**Comment by yairans** [comment]
@jessp01 do you mean `@HTML5_VER@` should be `@HTML5LIB_VERSION@` ? 
I took`@HTML5_VER@` from https://github.com/kaltura/platform-install-packages/search?q=%40HTML5_VER%40
Also you need to remove the versions from local.ini (which overrides the base.ini) if you have.

**Comment by jessp01** [comment]
Hi @yairans ,

Yes, unfortunately, in this company, we never adhere to one convention... Like I said, I'd rather we use:
```
@STUDIO_VERSION@
@STUDIO3_VERSION@
@HTML5LIB_VERSION@
```
Which is what the local.ini template I referenced uses. 
>  Also, you need to remove the versions from local.ini (which overrides the base.ini) if you have.
I'll remove the code that uncomments these and replaces the token, seeing how the new template makes it obsolete.

**Pull request merged**